### PR TITLE
Add debug info about variables to BIR

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Add debug info about variables to BIR
Related to #935

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/bir/bir.bal [hunk 1]
```diff
@@ -176,8 +176,37 @@ public type RegisterBase record {|
...
+# Represent scope inside the function body such as a while loop
+public type LexicalBlockScope readonly & record {|
+    LEXICAL_BLOCK_SCOPE_KIND kind = LEXICAL_BLOCK_SCOPE_KIND;
+    RegisterScope parentScope;
+    Position startPos;
+    Position endPos;
+|};
+
+# We don't need a reference to the actual function here since the scaffold will have it.
+public type FunctionScope readonly & record {|
```
Comment (jclark) [inline compiler/modules/bir/bir.bal]: Why do we need this?  Can't we use just use `LexicalBlockScope?` for scope?

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/bir/bir.bal (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/bir/bir.bal]: Fixed (Removed)

File: compiler/modules/bir/bir.bal [hunk 2]
```diff
@@ -176,8 +176,19 @@ public type RegisterBase record {|
 public type DeclRegister ParamRegister|VarRegister|FinalRegister;
 public type Register DeclRegister|NarrowRegister|TmpRegister|AssignTmpRegister;
 
+public type RegisterScope readonly & record {|
+    RegisterScope? parentScope;
+    Position startPos;
+    Position endPos;
+|};
+
+public function createRegisterScope(RegisterScope? parentScope, Position startPos, Position endPos) returns RegisterScope {
```
Comment (jclark) [inline compiler/modules/bir/bir.bal]: I don't think we need this.

File: compiler/modules/bir/bir.bal [hunk 3]
```diff
@@ -176,8 +176,19 @@ public type RegisterBase record {|
 public type DeclRegister ParamRegister|VarRegister|FinalRegister;
 public type Register DeclRegister|NarrowRegister|TmpRegister|AssignTmpRegister;
 
+public type RegisterScope readonly & record {|
+    RegisterScope? parentScope;
```
Comment (jclark) [inline compiler/modules/bir/bir.bal]: parentScope is unnecessarily wordy

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/bir/bir.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/bir/bir.bal]: fixed

File: compiler/modules/bir/bir.bal (hunk 3 above)
Comment (heshanpadmasiri) [reply compiler/modules/bir/bir.bal]: fixed

**Comment by jclark** [comment]
This needs to be updated to deal with the big patch from manu which just landed.

**Comment by heshanpadmasiri** [comment]
> This needs to be updated to deal with the big patch from manu which just landed.

fixed

**Comment by jclark** [comment]
LGTM

**Pull request merged**